lib/rspec/core/bisect/example_minimizer.rb

```diff
@@ -0,0 +1,130 @@
+      # repeatedly running different subsets of the suite.
+      class ExampleMinimizer
+        attr_reader :runner, :reporter, :all_example_ids, :failed_example_ids
+        attr_accessor :remaining_ids
+
+        def initialize(runner, reporter)
+          @runner   = runner
+          @reporter = reporter
+        end
+
+        def find_minimal_repro
+          prep
+
+          self.remaining_ids = non_failing_example_ids
+
+          each_bisect_round do |subsets|
+            ids_to_ignore = subsets.find do |ids|
+              get_expected_failures_for?(remaining_ids - ids)
+            end
+
+            next :done unless ids_to_ignore
+
+            self.remaining_ids -= ids_to_ignore
+            notify(:bisect_ignoring_ids, :ids_to_ignore => ids_to_ignore, :remaining_ids => remaining_ids)
+          end
+
+          currently_needed_ids
+        end
+
+        def currently_needed_ids
+          remaining_ids + failed_example_ids
+        end
+
+        def repro_command_for_currently_needed_ids
+          return runner.repro_command_from(currently_needed_ids) if remaining_ids
+          "(Not yet enough information to provide any repro command)"
+        end
+
+      private
+
+        def prep
+          notify(:bisect_starting, :original_cli_args => runner.original_cli_args)
+
+          _, duration = track_duration do
+            original_results    = runner.original_results
+            @all_example_ids    = original_results.all_example_ids
+            @failed_example_ids = original_results.failed_example_ids
+          end
+
+          if @failed_example_ids.empty?
+            raise BisectFailedError, "\n\nNo failures found. Bisect only works " \
+                  "in the presence of one or more failing examples."
+          else
+            notify(:bisect_original_run_complete, :failed_example_ids => failed_example_ids,
+                                                  :non_failing_example_ids => non_failing_example_ids,
+                                                  :duration => duration)
+          end
+        end
+
+        def non_failing_example_ids
+          @non_failing_example_ids ||= all_example_ids - failed_example_ids
+        end
+
+        def get_expected_failures_for?(ids)
+          ids_to_run = ids + failed_example_ids
+          notify(:bisect_individual_run_start, :command => runner.repro_command_from(ids_to_run))
+
+          results, duration = track_duration { runner.run(ids_to_run) }
+          notify(:bisect_individual_run_complete, :duration => duration, :results => results)
+
+          abort_if_ordering_inconsistent(results)
+          (failed_example_ids & results.failed_example_ids) == failed_example_ids
+        end
+
+        INFINITY = (1.0 / 0) # 1.8.7 doesn't define Float::INFINITY so we define our own...
+
+        def each_bisect_round(&block)
+          last_round, duration = track_duration do
+            1.upto(INFINITY) do |round|
```

Comment:
> makes me nervous because of infinite loop potential - non obvious that it will always terminate.

Proving that it will always terminate would require solving the halting problem, which can't be done, of course.

With a few reasonable assumptions, I think we can informally prove it, though.  Assuming:

* Each individual spec suite run terminates.  If it doesn't terminate...well, we can't do anything about that, and the non-termination has to do with the user's spec suite and not our use of `1.upto(infinity)` here.
* The number of total examples in the suite is constant for a given bisect run (e.g. the specs themselves aren't writing out new spec files adding new examples, etc)

Then:

* For any given bisect round, one of two things can happen:
  * It can find some example ids it can safely ignore, in which case it deletes them from `remaining_ids`.
  * It can find no example ids it can safely ignore, in which case it breaks from the loop with `:done`
* As long as the first case is hit, the number of remaining ids will decrease for each round
* Given we started with a finite number of examples, it'll have to terminate eventually

(I'm terrible at writing proofs but hopefully that makes sense).